# **Como usar NLTK en Google Colab**

In [1]:
#seleccionar download [d], luego descargar el recurso de nombre "book"
import nltk
nltk.download('cess_esp')  #descarga corpous contenido en la libreria NLTK. El texto con el cual vamos a trabajar

[nltk_data] Downloading package cess_esp to /root/nltk_data...
[nltk_data]   Unzipping corpora/cess_esp.zip.


True

# **Expresiones Regulares**


*   Constituyen un lenguaje estandarizado para definir cadenas de búsqueda de texto.
*   Libreria de operaciones con  expresiones regulares de Python [re](https://docs.python.org/3/library/re.html)
*   Reglas para escribir expresiones regulares [Wiki](https://es.wikipedia.org/wiki/Expresión_regular)



In [2]:
# spanish Corpus: https://mailman.uib.no/public/corpora/2007-October/005448.html
import re                                                                         #Libreria de expresiones Regulares
corpus = nltk.corpus.cess_esp.sents()                                             #Definios un objeto corpus
print(corpus)
print(len(corpus))                                                                #len() comando de python que calcula la longitud de una lista

[['El', 'grupo', 'estatal', 'Electricité_de_France', '-Fpa-', 'EDF', '-Fpt-', 'anunció', 'hoy', ',', 'jueves', ',', 'la', 'compra', 'del', '51_por_ciento', 'de', 'la', 'empresa', 'mexicana', 'Electricidad_Águila_de_Altamira', '-Fpa-', 'EAA', '-Fpt-', ',', 'creada', 'por', 'el', 'japonés', 'Mitsubishi_Corporation', 'para', 'poner_en_marcha', 'una', 'central', 'de', 'gas', 'de', '495', 'megavatios', '.'], ['Una', 'portavoz', 'de', 'EDF', 'explicó', 'a', 'EFE', 'que', 'el', 'proyecto', 'para', 'la', 'construcción', 'de', 'Altamira_2', ',', 'al', 'norte', 'de', 'Tampico', ',', 'prevé', 'la', 'utilización', 'de', 'gas', 'natural', 'como', 'combustible', 'principal', 'en', 'una', 'central', 'de', 'ciclo', 'combinado', 'que', 'debe', 'empezar', 'a', 'funcionar', 'en', 'mayo_del_2002', '.'], ...]
6030


En python los [] implican Listas. Como tengo dobles corchetes tengo una Lista que contiene varias Listas.

**Este corpus cada lista pertenece a titulares de Noticias en español y ya esta TOKENIZADO.**

El corpus esta compuesto por 6030 titulares de noticias en español

###Flatten

Flatten o aplanado implica que al tener una lista de listas, aplanamos todo en una sola Lista.

Es decir que originalmente tenemos una lista que contiene otras listas donde c/u de estas es un titular. Con Flaten obtenemos una sola lista donde estan todos los titulares aplanados

In [3]:
flatten = [w for l in corpus for w in l]   #ciclo de ciclo, por cada sublista hago una iteracion sobre las palabras (w) que estan en esa lista
print(flatten[:20])                        # [:20] primero 20 elementos
print(len(flatten))

['El', 'grupo', 'estatal', 'Electricité_de_France', '-Fpa-', 'EDF', '-Fpt-', 'anunció', 'hoy', ',', 'jueves', ',', 'la', 'compra', 'del', '51_por_ciento', 'de', 'la', 'empresa', 'mexicana']
192686


En vez de hacerlo con un ciclo de ciclo (flatten = [w for l in corpus for w in l]) podria realizar el flatten con el siguiente codigo

In [ ]:
flatten = []  # Inicializar una lista vacía

for l in corpus:  # Iterar sobre cada sublista en corpus
    for w in l:  # Iterar sobre cada elemento en la sublista l
        flatten.append(w)  # Agregar el elemento a la lista flatten

### **Estructura de la funcion re.search()**
```
# Determina si el patron de búsqueda p esta contenido en la cadena s
re.seach(p, s)
```



In [ ]:
# Meta-caracteres básicos
arr = [w for w in flatten if re.search('es', w)]  #Para cada objeto w (palabra) en flatten busca la expresion regular "es" y devuelve true si la encuentra, false en caso contrario
arr[:5]

['estatal', 'jueves', 'empresa', 'centrales', 'francesa']

In [ ]:
arr = [w for w in flatten if re.search('es$', w)]   # "es$" el $ indica que el patron tiene que estar al final de la palabra
arr[:5]

['jueves', 'centrales', 'millones', 'millones', 'dólares']

In [4]:
arr = [w for w in flatten if re.search('^es', w)]   # "^es" el ^ indica que el patron tiene que estar al principio de la palabra
arr[:5]

['estatal', 'es', 'esta', 'esta', 'eso']

In [ ]:
arr = [w for w in flatten if re.search('^..j..t..$', w)] # 3ra letra j, 5ta letra t contando desde el principio
arr

['tajantes']

##Rangos

Con corchetes en expresiones regulares indicamos un rango.
Por ejemplo  **^[ghi]**  indica que busque tokens o palabras (w) que empiezen con g, h o i

In [5]:
arr = [w for w in flatten if re.search('^[ghi]', w)]
arr[:5]

['grupo', 'hoy', 'gas', 'gas', 'intervendrá']

In [ ]:
#Rangos [a-z], [A-Z], [0-9]
arr = [w for w in flatten if re.search('^[ghi][mno][jlk][def]$', w)]
arr

['golf', 'golf']

### Clausuras

*  "*" se repite 0 o mas veces
+  "+" se repite 1 o mas veces

In [ ]:
#Clausuras *, * (Kleene closures)
arr = [w for w in flatten if re.search('^(no)*', w)]
arr[:10]

['El',
 'grupo',
 'estatal',
 'Electricité_de_France',
 '-Fpa-',
 'EDF',
 '-Fpt-',
 'anunció',
 'hoy',
 ',']

In [ ]:
arr = [w for w in flatten if re.search('(no)+', w)]
arr[:10]

['norte',
 'no',
 'no',
 'noche',
 'no',
 'no',
 'gobierno',
 'notificación',
 'Unión_Fenosa_Inversiones',
 'italiano']

# **Normalización de Texto** (como aplicación de las expresiones regulares)


## **Tokenización:** Es el proceso mediante el cual se sub-divide una cadena de texto en unidades linguísticas minimas (palabras)


In [ ]:
texto = """ Cuando sea el rey del mundo  (imaginaba él en su cabeza) no tendré que  preocuparme por estas bobadas.
            Era solo un niño de 7 años, pero pensaba que podría ser cualquier cosa que su imaginación le permitiera visualizar en su cabeza ..."""
print(texto)

 Cuando sea el rey del mundo  (imaginaba él en su cabeza) no tendré que  preocuparme por estas bobadas. 
            Era solo un niño de 7 años, pero pensaba que podría ser cualquier cosa que su imaginación le permitiera visualizar en su cabeza ...


In [ ]:
# Caso 1: tokenizacion más simple: por espacios vacios !
print(re.split(r' ', texto))

['', 'Cuando', 'sea', 'el', 'rey', 'del', 'mundo', '', '(imaginaba', 'él', 'en', 'su', 'cabeza)', 'no', 'tendré', 'que', '', 'preocuparme', 'por', 'estas', 'bobadas.', '\n', '', '', '', '', '', '', '', '', '', '', '', 'Era', 'solo', 'un', 'niño', 'de', '7', 'años,', 'pero', 'pensaba', 'que', 'podría', 'ser', 'cualquier', 'cosa', 'que', 'su', 'imaginación', 'le', 'permitiera', 'visualizar', 'en', 'su', 'cabeza', '...']


In [ ]:
# Caso 2: tokenización usando expresiones regulares
print(re.split(r'[ \t\n]+', texto))

['', 'Cuando', 'sea', 'el', 'rey', 'del', 'mundo', '(imaginaba', 'él', 'en', 'su', 'cabeza)', 'no', 'tendré', 'que', 'preocuparme', 'por', 'estas', 'bobadas.', 'Era', 'solo', 'un', 'niño', 'de', '7', 'años,', 'pero', 'pensaba', 'que', 'podría', 'ser', 'cualquier', 'cosa', 'que', 'su', 'imaginación', 'le', 'permitiera', 'visualizar', 'en', 'su', 'cabeza', '...']


In [ ]:
# RegEx reference: \W -> all characters other than letters, digits or underscore
print(re.split(r'[ \W\t\n]+', texto))

['', 'Cuando', 'sea', 'el', 'rey', 'del', 'mundo', 'imaginaba', 'él', 'en', 'su', 'cabeza', 'no', 'tendré', 'que', 'preocuparme', 'por', 'estas', 'bobadas', 'Era', 'solo', 'un', 'niño', 'de', '7', 'años', 'pero', 'pensaba', 'que', 'podría', 'ser', 'cualquier', 'cosa', 'que', 'su', 'imaginación', 'le', 'permitiera', 'visualizar', 'en', 'su', 'cabeza', '']


## **Tokenizador de NLTK**

In [ ]:
# nuestra antigua regex no funciona en este caso:
texto = 'En los E.U. esa postal vale $15.50 ...'
print(re.split(r'[ \W\t\n]+', texto))

['En', 'los', 'E', 'U', 'esa', 'postal', 'vale', '15', '50', '']


In [ ]:
pattern = r'''(?x)                 # set flag to allow verbose regexps
              (?:[A-Z]\.)+         # abbreviations, e.g. U.S.A.
              | \w+(?:-\w+)*       # words with optional internal hyphens
              | \$?\d+(?:\.\d+)?%? # currency and percentages, e.g. $12.40, 82%
              | \.\.\.             # ellipsis
              | [][.,;"'?():-_`]   # these are separate tokens; includes ], [
'''
nltk.regexp_tokenize(texto, pattern)

['En', 'los', 'E.U.', 'esa', 'postal', 'vale', '$15.50', '...']

## **Lematización:** Proceso para encontrar la raíz linguística de una palabra

*   Derivación (stemming) : lematización simple



In [ ]:
# Derivación simple
from nltk import word_tokenize
from nltk.stem.snowball import SnowballStemmer
SnowballStemmer.languages

('arabic',
 'danish',
 'dutch',
 'english',
 'finnish',
 'french',
 'german',
 'hungarian',
 'italian',
 'norwegian',
 'porter',
 'portuguese',
 'romanian',
 'russian',
 'spanish',
 'swedish')

In [ ]:
stem = SnowballStemmer('spanish')
stem.stem('trabajando')

'trabaj'

In [ ]:
# Lematización
from nltk.stem import WordNetLemmatizer
lemm = WordNetLemmatizer()

In [ ]:
lemm.lemmatize('trabajando')

'trabajando'